device/pcie/pci_device.cpp

```diff
@@ -37,6 +37,38 @@ static const uint32_t BH_BAR0_WC_MAPPING_SIZE = 188<<21;
 static const uint32_t BH_NOC_NODE_ID_OFFSET = 0x1FD04044;
 static const uint32_t GS_WH_ARC_SCRATCH_6_OFFSET = 0x1FF30078;
 
+static constexpr semver_t MINIMUM_SUPPORTED_KMD_VERSION{1, 27, 1};
+
+static semver_t read_kmd_version() {
+    static const std::string path = "/sys/module/tenstorrent/version";
+    std::ifstream file(path);
+
+    if (!file.is_open()) {
+       TT_THROW("Failed to open: {}; could not determine tenstorrent driver version", path);
```

Comment:
use log_assert.
And maybe switch to a warning?

Reply:
On second though, please DO switch this to a warning.
If for anycase something is different on some system, this will block any kind of workload and be a pain.

Reply:
I can make it return {0,0,0} here which will make it fail the check.  Are you suggesting that the check `if (kmd_version < MINIMUM_SUPPORTED_KMD_VERSION)` should be a warning?

If a system is missing `/sys/module/tenstorrent/version` then it simply does not have the driver installed.

Reply:
Yes, I'm saying both should be a warning, sorry for not being precise. Or if you make it return {0,0,0} here, than just the if (kmd_version < MINIMUM_SUPPORTED_KMD_VERSION) should throw a warning.
This has a potential to affect all existing systems (CI machine, Cloud machines, customer machines...), which might be functional today, but will make them non-functional due to this check.
There might be something custom on some system where this check might fail even though the driver is completely functional.
I'm not confident for the couple of us, without wider consensus, to make these hard constraints on our systems.
Logging a warning is always good though, since it could point in the right direction if something is failing.

Reply:
I rewrote most of this, please take another look.
For now the KMD version is simply logged.  Later on when UMD actually needs to impose a minimum version check, we'll have the infrastructure in place.

Reply:
If there is a minimum version needed for anything now, I don't mind printing a warning if version is not at least some predefined minimum version constant.